com.google.gson.JsonSyntaxException when parsing error response

Like in #103, we get an exception while parsing error output from the HTTP API when we pass an object when a String/number/... is expected.

```java
        final CMAField testField2 = new CMAField()
                .setName("Test Field 2")
                .setId("testField2")
                .setType(Symbol)
                .setRequired(true);

        final CMAContentType contentType = new CMAContentType()
                .setName(TEST_CONTENT_TYPE_FOR_UPDATE)
                .setId(TEST_CONTENT_TYPE_FOR_UPDATE)
                .setDescription("Description for test content type")
                .setFields(asList(testField2));

        final CMAContentType contentTypeWithAdditionalFields = CMA_CLIENT.contentTypes().create(SPACE_ID, contentType);
        CMA_CLIENT.contentTypes().publish(contentTypeWithAdditionalFields);
        
        CMA_CLIENT.entries().create(SPACE_ID, TEST_CONTENT_TYPE_FOR_UPDATE, new CMAEntry().setField("testField2", "de-DE", singletonList(0)));
```

The following exception seems to be raised when the CMA client tries to parse the response from the CMA endpoint that seems to reproduce the passed value.

```
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected a string but was BEGIN_OBJECT at line 11 column 19 path $.details.errors[0].value

	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:224)
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82)
	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
	at com.google.gson.Gson.fromJson(Gson.java:887)
	at com.google.gson.Gson.fromJson(Gson.java:852)
	at com.google.gson.Gson.fromJson(Gson.java:801)
	at com.google.gson.Gson.fromJson(Gson.java:773)
	at com.contentful.java.cma.model.CMAHttpException.<init>(CMAHttpException.java:244)
	at com.contentful.java.cma.interceptor.ErrorInterceptor.intercept(ErrorInterceptor.java:29)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at com.contentful.java.cma.interceptor.ContentTypeInterceptor.intercept(ContentTypeInterceptor.java:49)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at com.contentful.java.cma.interceptor.HeaderInterceptor.intercept(HeaderInterceptor.java:37)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at com.contentful.java.cma.interceptor.HeaderInterceptor.intercept(HeaderInterceptor.java:37)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at com.contentful.java.cma.interceptor.HeaderInterceptor.intercept(HeaderInterceptor.java:37)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185)
	at okhttp3.RealCall.execute(RealCall.java:69)
	at retrofit2.OkHttpCall.execute(OkHttpCall.java:180)
	at retrofit2.adapter.rxjava2.CallExecuteObservable.subscribeActual(CallExecuteObservable.java:41)
	at io.reactivex.Observable.subscribe(Observable.java:10838)
	at retrofit2.adapter.rxjava2.BodyObservable.subscribeActual(BodyObservable.java:34)
	at io.reactivex.Observable.subscribe(Observable.java:10838)
	at io.reactivex.internal.operators.flowable.FlowableFromObservable.subscribeActual(FlowableFromObservable.java:29)
	at io.reactivex.Flowable.subscribe(Flowable.java:12978)
	at io.reactivex.internal.operators.flowable.FlowableOnBackpressureLatest.subscribeActual(FlowableOnBackpressureLatest.java:32)
	at io.reactivex.Flowable.subscribe(Flowable.java:12978)
	at io.reactivex.Flowable.blockingFirst(Flowable.java:5221)
	at com.contentful.java.cma.ModuleEntries.create(ModuleEntries.java:88)
	at de.henka.shop.contentful.setup.content.type.ContentTypeUpdateTest.updateExistingFieldWithEntries(ContentTypeUpdateTest.java:62)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.lang.IllegalStateException: Expected a string but was BEGIN_OBJECT at line 11 column 19 path $.details.errors[0].value
	at com.google.gson.stream.JsonReader.nextString(JsonReader.java:825)
	at com.google.gson.internal.bind.TypeAdapters$16.read(TypeAdapters.java:418)
	at com.google.gson.internal.bind.TypeAdapters$16.read(TypeAdapters.java:406)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
	... 65 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,9 @@
 All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
+## Version [2.2.3] - 2018-03-08
+- Fix: Parsing error in error response
+
 ## Version [2.2.2] - 2018-02-19
 - Fix: make creating auto save webhooks work.
 - Fix: add details to contentful error.
@@ -121,7 +124,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ## Version 0.9.0 - 2014-11-18
 Initial release.
 
-[unreleased]: https://github.com/contentful/contentful-management.java/compare/cma-sdk-2.2.2...HEAD
+[unreleased]: https://github.com/contentful/contentful-management.java/compare/cma-sdk-2.2.3...HEAD
+[2.2.3]: https://github.com/contentful/contentful-management.java/compare/cma-sdk-2.2.2...cma-sdk-2.2.3
 [2.2.2]: https://github.com/contentful/contentful-management.java/compare/cma-sdk-2.2.1...cma-sdk-2.2.2
 [2.2.1]: https://github.com/contentful/contentful-management.java/compare/cma-sdk-2.2.0...cma-sdk-2.2.1
 [2.2.0]: https://github.com/contentful/contentful-management.java/compare/cma-sdk-2.1.0...cma-sdk-2.2.0

--- a/README.md
+++ b/README.md
@@ -16,12 +16,12 @@ Grab via Maven:
 <dependency>
   <groupId>com.contentful.java</groupId>
   <artifactId>cma-sdk</artifactId>
-  <version>2.2.2</version>
+  <version>2.2.3</version>
 </dependency>
 ```
 or Gradle:
 ```groovy
-compile 'com.contentful.java:cma-sdk:2.2.2'
+compile 'com.contentful.java:cma-sdk:2.2.3'
 ```
 
 The SDK requires at minimum Java 6 or Android 2.3.

--- a/src/main/java/com/contentful/java/cma/model/CMAHttpException.java
+++ b/src/main/java/com/contentful/java/cma/model/CMAHttpException.java
@@ -66,8 +66,9 @@ public class CMAHttpException extends RuntimeException {
         String details;
         String type;
         String filter;
-        String value;
+        Object value;
         Object path;
+        List<String> expected;
 
         /**
          * @return the name of this error.
@@ -100,8 +101,15 @@ public class CMAHttpException extends RuntimeException {
         /**
          * @return the value triggering this error.
          */
-        public String getValue() {
+        public Object getValue() {
           return value;
+        }
+
+        /**
+         * @return an invalid type returns an expectation.
+         */
+        public List<String> getExpected() {
+          return expected;
         }
 
         /**

--- a/src/test/bash/content_type_error_field_type_invalid.sh
+++ b/src/test/bash/content_type_error_field_type_invalid.sh
@@ -1,0 +1,12 @@
+#!/usr/bin/env bash
+
+output="$(basename $(echo $0) | sed 's#sh$#json#g' | sed 's#^#../resources/#g')"
+
+curl --verbose \
+    -X POST \
+    -H 'Content-Type: application/vnd.contentful.management.v1+json' \
+    -H "X-Contentful-Content-Type: TESTID" \
+    -H 'Authorization: Bearer '$CMA_TOKEN  \
+    -d '{"sys":{"id":"InvalidContentType"},"fields":[{"name":"name","id":"testid","type":"null"}],"name":"Invalid Content Type"}' \
+    "https://api.contentful.com/spaces/$SPACE_ID/content_types/" \
+   | tee ${output}

--- a/src/test/bash/entry_create_error_array_type_value.sh
+++ b/src/test/bash/entry_create_error_array_type_value.sh
@@ -1,0 +1,12 @@
+#!/usr/bin/env bash
+
+output="$(basename $(echo $0) | sed 's#sh$#json#g' | sed 's#^#../resources/#g')"
+
+curl --verbose \
+    -X POST \
+    -H 'Content-Type: application/vnd.contentful.management.v1+json' \
+    -H "X-Contentful-Content-Type: TESTID" \
+    -H 'Authorization: Bearer '$CMA_TOKEN  \
+    -d '{"fields":{"testField2":{"en-US":[2]}}}' \
+    'https://api.contentful.com/spaces/'$SPACE_ID'/entries/' \
+   | tee ${output}

--- a/src/test/kotlin/com/contentful/java/cma/ContentTypeTests.kt
+++ b/src/test/kotlin/com/contentful/java/cma/ContentTypeTests.kt
@@ -22,10 +22,12 @@ import com.contentful.java.cma.lib.TestCallback
 import com.contentful.java.cma.lib.TestUtils
 import com.contentful.java.cma.model.CMAContentType
 import com.contentful.java.cma.model.CMAField
+import com.contentful.java.cma.model.CMAHttpException
 import com.contentful.java.cma.model.CMAType
 import okhttp3.HttpUrl
 import okhttp3.mockwebserver.MockResponse
 import java.io.IOException
+import java.util.Collections
 import kotlin.test.assertEquals
 import kotlin.test.assertFalse
 import kotlin.test.assertNotNull
@@ -419,5 +421,27 @@ class ContentTypeTests : BaseTest() {
         assertEquals("GET", recordedRequest.method)
         assertEquals("/spaces/spaceId/content_types/contentTypeId/snapshots/snapShotId",
                 recordedRequest.path)
+    }
+
+    @test(expected = CMAHttpException::class)
+    fun testFailureOnCreateContenTypeWithWrongFieldType() {
+        val responseBody = TestUtils.fileToString("content_type_error_field_type_invalid.json")
+        server!!.enqueue(MockResponse().setResponseCode(422).setBody(responseBody))
+
+        val fields = Collections.singletonList(CMAField().setName("name").setId("testid").setType(null).setRequired(true))
+
+        val contentType = CMAContentType()
+                .setId("entryId")
+                .setSpaceId("spaceId")
+                .setFields(fields)
+
+        try {
+            client!!.contentTypes().create("spaceID", contentType)
+        } catch (e: CMAHttpException) {
+            assertEquals("null", e.errorBody.details.errors[0].value)
+            assertEquals("[Array, Boolean, Date, Integer, Number, Object, Symbol, Link, Location, Text]",
+                    e.errorBody.details.errors[0].expected.toString())
+            throw e
+        }
     }
 }

--- a/src/test/kotlin/com/contentful/java/cma/EntryTests.kt
+++ b/src/test/kotlin/com/contentful/java/cma/EntryTests.kt
@@ -20,10 +20,12 @@ import com.contentful.java.cma.lib.ModuleTestUtils
 import com.contentful.java.cma.lib.TestCallback
 import com.contentful.java.cma.lib.TestUtils
 import com.contentful.java.cma.model.CMAEntry
+import com.contentful.java.cma.model.CMAHttpException
 import com.contentful.java.cma.model.CMAType
 import okhttp3.HttpUrl
 import okhttp3.mockwebserver.MockResponse
 import java.io.IOException
+import java.util.*
 import kotlin.test.*
 import org.junit.Test as test
 
@@ -308,6 +310,25 @@ class EntryTests : BaseTest() {
         }
     }
 
+    @test(expected = CMAHttpException::class)
+    fun testFailureOnCreateEntryWithWrongFieldType() {
+        val responseBody = TestUtils.fileToString("entry_create_error_array_type_value.json")
+        server!!.enqueue(MockResponse().setResponseCode(422).setBody(responseBody))
+
+        val entry = CMAEntry()
+                .setId("entryId")
+                .setSpaceId("spaceId")
+                .setField("testField2", "en-US", Collections.singletonList(""))
+
+        try {
+            client!!.entries().create("spaceID", "contenttype", entry)
+        } catch (e: CMAHttpException) {
+            assertEquals(
+                    "The type of \"value\" is incorrect, expected type: Symbol",
+                    e.errorBody.details.errors[0].details)
+            throw e
+        }
+    }
     @test
     fun testChainingInterface() {
         val entry = CMAEntry()

--- a/src/test/resources/content_type_error_field_type_invalid.json
+++ b/src/test/resources/content_type_error_field_type_invalid.json
@@ -1,0 +1,33 @@
+{
+  "sys": {
+    "type": "Error",
+    "id": "ValidationFailed"
+  },
+  "message": "Validation error",
+  "details": {
+    "errors": [
+      {
+        "name": "in",
+        "path": [
+          "fields",
+          0,
+          "type"
+        ],
+        "value": "null",
+        "expected": [
+          "Array",
+          "Boolean",
+          "Date",
+          "Integer",
+          "Number",
+          "Object",
+          "Symbol",
+          "Link",
+          "Location",
+          "Text"
+        ]
+      }
+    ]
+  },
+  "requestId": "ca5845b2efc9cebbc417e25dec1d3277"
+}

--- a/src/test/resources/entry_create_error_array_type_value.json
+++ b/src/test/resources/entry_create_error_array_type_value.json
@@ -1,0 +1,25 @@
+{
+  "sys": {
+    "type": "Error",
+    "id": "InvalidEntry"
+  },
+  "message": "Validation error",
+  "details": {
+    "errors": [
+      {
+        "name": "type",
+        "value": [
+          2
+        ],
+        "type": "Symbol",
+        "details": "The type of \"value\" is incorrect, expected type: Symbol",
+        "path": [
+          "fields",
+          "testField2",
+          "en-US"
+        ]
+      }
+    ]
+  },
+  "requestId": "fa607c736302207868b25e97c18ad3dd"
+}

